dap-dlv-go.el

```diff
@@ -0,0 +1,132 @@
+               ("auto"
+		(dap-dlv-go--populate-auto-args conf))
+               ("debug"
+		(dap--put-if-absent
+		 conf :program (f-dirname (buffer-file-name))))
+               ("exec"
+		(dap--put-if-absent
+		 conf :program
+		 (f-expand (read-file-name "enter path to executable: "))))
+               ("remote"
+		(dap--put-if-absent conf :host (read-string "enter host: " "127.0.0.1"))
+		(dap--put-if-absent conf :debugPort
+				    (string-to-number (read-string "enter port: " "2345"))))
+               ("local"
+                (dap--put-if-absent conf :cwd (f-dirname (buffer-file-name)))
+                (dap--put-if-absent
+		 conf :processId (string-to-number (read-string "enter pid: " "2345"))))))
+
+  (let ((debug-port (if (string= (plist-get conf :mode)
+				 "remote")
+			(plist-get conf :debugPort)
+		      (dap--find-available-port))))
+    (dap--put-if-absent conf :host "localhost")
+    (when (not (string= "remote" (plist-get conf :mode)))
+      (plist-put
+       conf :program-to-start
+       (format "%s dap --listen 127.0.0.1:%s %s" dap-dlv-go-delve-path debug-port dap-dlv-go-extra-args)))
+    (plist-put conf :debugServer debug-port))
+
+  (if (stringp (plist-get conf :args)) (plist-put conf :args (split-string (plist-get conf :args))) ())
+  
+  (-> conf
+      (dap--put-if-absent :dlvToolPath dap-dlv-go-delve-path)
+      
+      (dap--put-if-absent :type "go")
+      (dap--put-if-absent :name "Go Dlv Debug")))
+
+(defun dap-dlv-go--populate-auto-args (conf)
+  "Populate auto arguments according to CONF."
+  (dap--put-if-absent conf :program (buffer-file-name))
+
+  (if (string-suffix-p "_test.go" (buffer-file-name))
+      (plist-put conf :mode "test")
+    (plist-put conf :mode "debug")))
+
+(dap-register-debug-provider "go" 'dap-dlv-go--populate-default-args)
+
+(dap-register-debug-template "Go Dlv Launch File Configuration"
+                             (list :type "go"
+                                   :request "launch"
+                                   :name "Launch File"
+                                   :mode "auto"
+                                   :program nil
+                                   :buildFlags nil
+                                   :args nil
+                                   :env nil))
+
+(dap-register-debug-template "Go Dlv Attach Configuration"
+                             (list :type "go"
+                                   :request "attach"
+                                   :name "Attach to running process"
+                                   :mode "auto"
+                                   :env nil))
+
+(dap-register-debug-template "Go Dlv Launch Executable Configuration"
+                             (list :type "go"
+                                   :request "exec"
+                                   :name "Launch Executable"
+                                   :mode "exec"
+                                   :program nil
+                                   :buildFlags nil
```

Comment:
I see you have removed some of the args that don't apply to some of the modes, but not all. Not sure if this was intentional or we need better documentation. Just in case here it is:  https://github.com/go-delve/delve/pull/2991. Delve won't really care if you give it extra flags, but if this shapes what users see they can configure, it might need more massaging.

Reply:
Thank you, now it's understandable. It was not clear for me from godoc.